lib/ansible/module_utils/api.py

```diff
@@ -114,3 +115,69 @@ def retried(*args, **kwargs):
 
         return retried
     return wrapper
+
+
+def generate_jittered_backoff(retries=10, delay_base=3, delay_threshold=60):
+    """The "Full Jitter" backoff strategy.
+
+    Ref: https://www.awsarchitectureblog.com/2015/03/backoff.html
+
+    :param retries: The number of delays to generate.
+    :param delay_base: The base time in seconds used to calculate the exponential backoff.
+    :param delay_threshold: The maximum time in seconds for any delay.
+    """
+    for retry in range(0, retries):
+        yield random.randint(0, min(delay_threshold, delay_base * 2 ** retry))
+
+
+def retry_never(exception_or_result):
+    return False
+
+
+def do_nothing(delay, exception=None, result=None):
+    return
+
+
+def retry_with_delays_and_condition(backoff_iterator, do_on_retry=None, should_retry_error=None, should_retry_result=None):
+    """Generic retry decorator.
+
+    :param backoff_iterator: An iterable of delays in seconds.
+    :param do_on_retry: A callable that takes the current delay, the exception/None, the result/None, and executes before the sleep (returns None).
+    :param should_retry_error: A callable that takes an exception of the decorated function and decides whether to retry or not (returns a bool).
+    :param should_retry_result: A callable that accepts the decorated function return value and decides whether or not to retry (returns a bool).
+    """
+    if do_on_retry is None:
+        do_on_retry = do_nothing
+    if should_retry_error is None:
+        should_retry_error = retry_never
+    if should_retry_result is None:
+        should_retry_result = retry_never
+
+    def function_wrapper(function):
```

Comment:
I overlooked this small detail — the current implementation doesn't copy over the underlying function's metadata:
```suggestion
    @functools.wraps(function)  # FIXME: use wrapt once it's in
    def function_wrapper(function):
```
This is necessary for the discovery in REPL and docs generation tools to pick up the proper `__doc__` and the argspec.

Also, there's talk, https://wrapt.rtfd.io is going to become a runtime dependency for the sake of object proxies feature in 2.12. So in the future, it's probably best to use it instead (AFAIK it does a better job taking care of the decorator corner cases).

Reply:
Great catch, thanks.